File: 0746-min-cost-climbing-stairs/0746-min-cost-climbing-stairs.py
class Solution:
    def minCostClimbingStairs(self, cost: List[int]) -> int:
        memo={}
        def dp(i):
            if i>=len(cost):
                return 0
            if i in memo:
                return memo[i]
            res=min(dp(i+1),dp(i+2))+cost[i]
            
            memo[i]=res
            return res
        return min(dp(0),dp(1))

File: 0746-min-cost-climbing-stairs/0746-min-cost-climbing-stairs.js
/**
 * @param {number[]} cost
 * @return {number}
 */
var minCostClimbingStairs = function(cost) {
    memo={}
    
    var dp=function(i){
        
        if(i>=cost.length){
            return 0
        }
        
        if(i in memo){
            return memo[i]
        }
        
        let result=Math.min(dp(i+1),dp(i+2))+cost[i]
        memo[i]=result
        return result
    }
    return Math.min(dp(0),dp(1))
};
